src/libexpr/primops/fetchTree.cc

```diff
@@ -248,7 +241,7 @@ static void fetch(EvalState & state, const Pos & pos, Value * * args, Value & v,
 
         auto validPaths = state.store->queryValidPaths({expectedPath}, NoSubstitute);
```

Comment:
Stupid question, but would passing `Substitute` here fix #4313 ?

Reply:
Great question.
It would, but there would also be a performance cost.

Remote path querying has a significant latency, that I don't think we want to incur on every fetch, many of which will not be substitutable anyway. Ideally we could perform the remote path info query while the fetcher runs, and make some clever decisions regarding cancellation. That's not something I want to dive into, especially considering the non-zero risk of corrupting local files when cancelling the fetcher.

Instead I figured the next step could be to try substitution when the fetching fails, which also solves #4313, without performance impact. So now the question is; do you like big or small PRs?

Reply:
Can we add a FIXME here linking #4313? Fine if just in the issue we add the details about latency.

Reply:
I think substituting is fine since the binary cache substituter caches negative lookup results. I.e. we can call `ensurePath` here instead of `queryValidPaths`.

BTW, why `queryValidPaths` instead of `store->isValidPath(expectedPath)`?

Reply:
I don't see how negative caching helps here. When a path is not in the cache, it will be fetched, after which it won't have to be looked up again. I suppose it might be useful to start an asynchronous query and then rely on the cache, but then the real benefit is in positive caching anyway.

> why `queryValidPaths`

I glossed over it. Maybe it's because `valid` makes me uneasy. Validity just doesn't seem as strong a property as existence. Yet, I've found `queryValidPaths` so maybe I should stop analyzing myself :sweat_smile:

Reply:
You're right, negative caching doesn't matter. But it also means that the overhead of substitution is a non-issue since it will only happen once.